includes/admin/home/tasks/task/class-sensei-home-task-publish-first-course.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * File containing the Sensei_Home_Task_Publish_First_Course class.
+ *
+ * @package sensei-lms
+ * @since $$next-version$$
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Sensei_Home_Task_Publish_First_Course class.
+ *
+ * @since $$next-version$$
+ */
+class Sensei_Home_Task_Publish_First_Course implements Sensei_Home_Task {
+	/**
+	 * The ID for the task.
+	 *
+	 * @return string
+	 */
+	public static function get_id(): string {
+		return 'publish-first-course';
+	}
+
+	/**
+	 * Number used to sort in frontend.
+	 *
+	 * @return int
+	 */
+	public function get_priority(): int {
+		return 600;
+	}
+
+	/**
+	 * Task title.
+	 *
+	 * @return string
+	 */
+	public function get_title(): string {
+		return __( 'Publish your first Course', 'sensei-lms' );
+	}
+
+	/**
+	 * Task url.
+	 *
+	 * @return string
+	 */
+	public function get_url(): ?string {
+		$result = $this->load_result();
+		if ( null === $result ) {
+			// If we only have the demo course (or no course at all), let's point the user to the new course page.
+			return admin_url( 'post-new.php?post_type=course' );
+		}
+		if ( '1' === $result->published ) {
+			// We shouldn't return a URL if the task is completed.
+			return null;
+		}
+		return get_edit_post_link( $result->ID, 'api' );
+	}
+
+	/**
+	 * Whether the task is completed or not.
+	 *
+	 * @return bool
+	 */
+	public function is_completed(): bool {
+		$result = $this->load_result();
+		return null !== $result && '1' === $result->published;
+	}
+
+	/**
+	 * Searches on the database for the latest course edited and/or published, so we can detect if
+	 * the edit URL for that course and if the task was published with the same query.
+	 *
+	 * @return stdClass|null The row containing the result, or null if there's no course that isn't the sample course.
+	 */
+	protected function load_result() {
+		global $wpdb;
+		$cache_key   = 'home/tasks/publish-first-course';
+		$cache_group = 'sensei/temporary';
+		$result      = wp_cache_get( $cache_key, $cache_group );
+		if ( false === $result ) {
+			$prefix = $wpdb->esc_like( Sensei_Data_Port_Manager::SAMPLE_COURSE_SLUG );
+			// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery -- Safe-ish query.
+			$result = $wpdb->get_row( $wpdb->prepare( "SELECT ID, post_status='publish' as published FROM {$wpdb->posts} WHERE post_type='course' AND post_name NOT LIKE %s ORDER BY published DESC, ID DESC LIMIT 1", "{$prefix}%" ) );
+			wp_cache_set( $cache_key, $result, $cache_group );
```

Comment:
I know we'd have to ignore the PHPCS warning, but I wonder if it is worth removing caching here. We're caching without any expiration. If there is any course (including trash), we're caching the object (without possibly `published === 0`) "forever". If there are no courses, `get_row` returns `null`. It is up to cache providers how to handle `null`, but it is possible that it will get stored and then `false !== $result` so it would also be returned "forever."

I think we also should probably do some filtering `post_status IN ('publish', 'draft')` so we never return trashed courses. Lastly, when all courses are in draft, do we want the latest created course or the _first_ course? If first, we might want to do `ID ASC`, I think.

Reply:
Hey there!

> I know we'd have to ignore the PHPCS warning, but I wonder if it is worth removing caching here. We're caching without any expiration

As explained in https://github.com/Automattic/sensei/pull/5854#discussion_r995165448 - this shouldn't cache forever, given that I'm using `sensei/temporary` as cache group, and that cache group is added as non persistent (using this [WP function](https://developer.wordpress.org/reference/functions/wp_cache_add_non_persistent_groups/)) here: https://github.com/Automattic/sensei/blob/c50c2339b57308fa6f6e69a0892eedc361bdd4f1/includes/class-sensei.php#L350 

With that said:

> If there is any course (including trash)

As explained on the PR above, I created https://github.com/Automattic/sensei/issues/5934 to track this.

> we're caching the object (without possibly published === 0) "forever". If there are no courses, get_row returns null. It is up to cache providers how to handle null, but it is possible that it will get stored and then false !== $result so it would also be returned "forever."

I'm not sure if there's any effect to this given what I explained above. 

> Lastly, when all courses are in draft, do we want the latest created course or the first course? If first, we might want to do `ID ASC`, I think.

Good catch! I originally implemented `ID DESC` to look for the last course, as I assume it should be the very latest course the user worked on. However, now I'm not 100% sure. I feel like both should work well, but I think I'll change to `ID ASC` just to avoid surprises and to maintain some consistency.

Reply:
Hey, fixed in [d758783](https://github.com/Automattic/sensei/pull/5888/commits/d75878341a70314a52f38a6e34133d239362b14d), [8f6ebb2](https://github.com/Automattic/sensei/pull/5888/commits/8f6ebb26faee378845af466e394bab8a3b0ab902) and 9e2f753938ab21e569546fc10536b47bcd7c2b32 

Hope it fixes all the points you mentioned, but let me know if it doesn't. :)